clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1491,20 +1495,25 @@ public void setPartition(int partition) {
 
             if (log.isTraceEnabled()) {
```

Comment:
A couple of lines above we use a language level assert. In Kafka, we typically use asset like methods like the Objects class since the language level asserts are disabled by default.

Reply:
The intention of an assert is to run in tests, but be disabled in prod, so if my understanding is correct, this is the proper usage.

Reply:
That's not considered best practice in Java and we don't do it in Kafka. What would be the reason to run it in tests only?

Reply:
It seems we have been mostly using Objects.requireNonNull for null assertion in our code. It doesn't seem to add too much overhead and helps identify issues in production early on. For consistency, perhaps we could use Objects.requireNonNull instead of assert.

@ijuma : What do you recommend that we use for assertions like `assert partitionInfo == stickyPartitionInfo.get()`?

Reply:
Correct @junrao, `Objects.requireNonNull` would be the recommended way to assert non null. The reference equality check is less common, we could add our own utility method in Utils for that or inline it.

The main thing is to get the appropriate signal if this happens in prod when the cost is low (both examples would be in that category).

Reply:
Ok, so what's the suggestion in this change?  Should I leave the code as is or remove assert?  Creating a new utility seems to be out of scope for this change.  We could have an offline discussion about asserts, I would be happy to see them used more often in Kafka.

Reply:
For the line in this method, you could do something like:

```java
if (partition < 0)
  throw new IllegalArgumentException("partition should be positive, but it was " + partition):
```

Which is more informative and idiomatic and checks the more general case that we expect partitions to be positive. But I see that we have sprinkled the same check in other methods. So, having a `assertPartitionIsPositive` would probably be a better approach. In any case, since this code was introduced in a different change, we can file a JIRA and do it as a separate PR.

I am happy to discuss more, but we should be clear about terminology. Language level asserts in Java aren't used much. Checking preconditions through API boundaries is useful. Within a given boundary, it's best to use the type system to avoid having noise all over the code.

Reply:
KAFKA-14085